File: frontend/src/types/scheme.ts
export interface Scheme {
  id: number;
  name: string;
  short_description: string;
  department: string;
  state: string;
  scheme_type: string;
  benefits: string;
}

export interface EligibilityCriterion {
  criterion_type: string;
  field_name: string;
  operator: string;
  expected_value: string;
  description: string | null;
}

export interface SchemeDetail extends Scheme {
  detailed_description: string;
  application_process: string;
  official_website: string | null;
  eligibility_criteria: EligibilityCriterion[];
}

export interface SchemeSearchResult {
  schemes: Scheme[];
  total: number;
}

export interface SchemeEligibilityResult {
  scheme: SchemeDetail;
  eligible: boolean;
  reasons: string[];
}

export type MatchStatus = "eligible" | "partial" | "no_match";

export interface SchemeRecommendation {
  scheme: Scheme;
  relevance_score: number;
  eligible: boolean;
  summary: string;
  factors: string[];
  match_status: MatchStatus;
}

export interface SchemeRecommendationsResult {
  recommendations: SchemeRecommendation[];
  total: number;
}

export interface SchemeFilters {
  search?: string;
  state?: string;
  scheme_type?: string;
}

File: backend/app/schemas/scheme.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class EligibilityCriterionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    criterion_type: str
    field_name: str
    operator: str
    expected_value: str
    description: str | None = None


class SchemeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    short_description: str
    department: str
    state: str
    scheme_type: str
    benefits: str


class SchemeDetailResponse(SchemeResponse):
    detailed_description: str
    application_process: str
    official_website: str | None = None
    eligibility_criteria: list[EligibilityCriterionResponse] = Field(default_factory=list)


class SchemeSearchResponse(BaseModel):
    schemes: list[SchemeResponse]
    total: int


class SchemeEligibilityResponse(BaseModel):
    scheme: SchemeDetailResponse
    eligible: bool
    reasons: list[str]


class SchemeRecommendationResponse(BaseModel):
    scheme: SchemeResponse
    relevance_score: int = Field(ge=0, le=100)
    eligible: bool
    summary: str
    factors: list[str] = Field(default_factory=list)
    match_status: Literal["eligible", "partial", "no_match"] = "no_match"


class SchemeRecommendationsListResponse(BaseModel):
    recommendations: list[SchemeRecommendationResponse]
    total: int
